src/components/validation/runner/telemetry.rs

```diff
@@ -113,28 +115,40 @@ impl Telemetry {
                         let mut events_seen = 0;
                         let current_time = chrono::Utc::now();
 
+                        let timeout = tokio::time::sleep(Duration::from_secs(5));
+                        tokio::pin!(timeout);
+
                         loop {
-                            match &rx.recv().await {
-                                None => break 'outer,
-                                Some(telemetry_event) => {
-                                    telemetry_events.push(telemetry_event.clone());
-                                    if let Event::Metric(metric) = telemetry_event {
-                                        if let Some(tags) = metric.tags() {
-                                            if metric.name() == INTERNAL_METRICS_SHUTDOWN_EVENT &&
-                                            tags.get("component_name") == Some(INTERNAL_LOGS_KEY) &&
-                                            metric.data().timestamp().unwrap() > &current_time {
-                                                debug!("Telemetry: processed one component_received_events_total event.");
-
-                                                events_seen += 1;
-                                                if events_seen == 2 {
-                                                    break 'outer;
+                            select! {
+                                d = rx.recv() => {
+                                    match d {
+                                        None => break,
+                                        Some(telemetry_event) => {
+                                            telemetry_events.push(telemetry_event.clone());
+                                            if let Event::Metric(metric) = telemetry_event {
+                                                if let Some(tags) = metric.tags() {
+                                                    if metric.name() == INTERNAL_METRICS_SHUTDOWN_EVENT &&
+                                                    tags.get("component_name") == Some(INTERNAL_LOGS_KEY) &&
+                                                    metric.data().timestamp().unwrap() > &current_time {
+                                                        debug!("Telemetry: processed one component_received_events_total event.");
+
+                                                        events_seen += 1;
```

Comment:
I think this logic (which isn't yours, but, now that I'm looking closer...) is buggered.

Since this loop is receiving a _single_ event at a time, we're only waiting for 3 events at most, which has nothing to do, necessarily, with how many "ticks" of the `internal_metrics` source have occurred.

What we probably want to do is change the `EventForwardService` to send an entire batch all at once. This way, based on the batch settings of the telemetry sink -- which should be able to hold all metrics emitted in a single tick by the `internal_metrics` source -- we'd get all of those metrics each time it ticked.

Then we could properly build up an invariant here of knowing that for each time we received an item, that item is actually the full set of metrics spit out by the `internal_metrics` source in a single tick.

Reply:
Hmm based on this, it sounds like we can change the `SHUTDOWN_TICKS` to go back to 2 with that correction.

Reply:
I took a swing at that in [0577ee6](https://github.com/vectordotdev/vector/pull/17956/commits/0577ee62f4b7b06d4d1f07e0adcdf25715a89a87)

It seems to be doing fine so far with going back to 2 ticks. I'll keep it running for a while.